configurator/sshkey/sshkey.go

```diff
@@ -0,0 +1,109 @@
+package sshkey
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/sys/unix"
+
+	"github.com/percona/pmm-manage/configurator/config"
+)
+
+func Init(c config.PMMConfig) Handler {
+	return Handler{
+		KeyPath:  c.SSHKeyPath,
+		KeyOwner: c.SSHKeyOwner,
+	}
+}
+
+func (c *Handler) RunSSHKeyChecks() {
+	sshKeyUser, err := user.Lookup(c.KeyOwner)
+	if err != nil {
+		log.Fatal(err)
+	}
+	if c.KeyPath == "" {
+		c.KeyPath = path.Join(sshKeyUser.HomeDir, ".ssh/authorized_keys")
+	}
+
+	sshKeyDir := filepath.Dir(c.KeyPath)
+	logger := log.WithField("dir", sshKeyDir)
+	if dir, err := os.Stat(sshKeyDir); err != nil || !dir.IsDir() {
+		if err := os.MkdirAll(sshKeyDir+"/", 0700); err != nil {
+			logger.WithFields(log.Fields{
+				"error": err,
+			}).Fatal("Cannot create ssh directory")
+		}
+		uid, _ := strconv.Atoi(sshKeyUser.Uid)
+		gid, _ := strconv.Atoi(sshKeyUser.Gid)
+		if err := os.Chown(sshKeyDir, uid, gid); err != nil {
+			logger.WithFields(log.Fields{
+				"error": err,
+			}).Fatal("Cannot change owner of ssh directory")
+		}
+	}
+	if err := unix.Access(sshKeyDir, unix.W_OK); err != nil {
+		logger.WithFields(log.Fields{
+			"error": err,
+		}).Fatal("Cannot write to ssh directory")
+	}
+}
+
+func parse(authorizedKey []byte) (Key, error) {
+	pubKey, comment, _, _, err := ssh.ParseAuthorizedKey(authorizedKey)
+	if err != nil {
+		return Key{}, err
+	}
+	return Key{
+		Type:        pubKey.Type(),
+		Comment:     comment,
+		Fingerprint: ssh.FingerprintSHA256(pubKey),
+	}, err
+}
+
+func (c *Handler) Read() (Key, string, error) {
+	authorizedKey, err := ioutil.ReadFile(c.KeyPath)
+	if err != nil {
+		return Key{}, "Cannot read ssh key", err
+	}
+	sshKey, err := parse(authorizedKey)
+	if err != nil {
+		return sshKey, "Cannot parse ssh key", err
+	}
+	return sshKey, "success", nil
+}
+
+func (c *Handler) Write(body io.ReadCloser) (Key, string, error) {
```

Comment:
When you return a struct by value, every field is pushed to the stack, it can be slow for a large struct. When you return a pointer, you push only a pointer. That's the original reason. After some time that rule had become an idiomatic way to return a struct of any size. There are two exceptions: when you return a `struct{}` (it takes zero bytes, can be used as a signal of some kind) and when you have a specific reason to return a copy by value (and then it should be in the comment). Both exceptions don't apply there.